plugin/src/main/java/org/opensearch/ml/model/MLModelAutoReDeployer.java

```diff
@@ -0,0 +1,511 @@
+                Map<String, Object> sourceAsMap = getReTryTimesResponse.getHits().getHits()[0].getSourceAsMap();
+                retryTimes = (Integer) sourceAsMap.get(MODEL_DEPLOY_RETRY_TIMES_FIELD);
+            }
+
+            // According to the node id to get deploy retry times, if more than the max deploy retry
+            // times, don't need to retry
+            // that the number of unsuccessful deploy has reached the maximum number of times, do not
+            // need to deploy
+            if (retryTimes > autoDeployMaxRetryTimes) {
+                log.info("Node: {} has reached to the max retry limit, failed to deploy models", localNodeId);
+                return;
+            }
+
+            String[] modelIds = modelIdAndNodeIds.keySet().toArray(new String[] {});
+            final int finalRetryTimes = retryTimes + 1;
+            List<ActionListener<MLDeployModelResponse>> listeners = Arrays
+                .stream(modelIds)
+                .map(e -> ActionListener.<MLDeployModelResponse>wrap(response -> {
+                    String[] nodeIds = modelIdAndNodeIds.get(e).toArray(new String[] {});
+                    log.info("the model {} is auto deploying under the node ids: {} ", e, nodeIds);
+                }, exception -> {
+                    String[] nodeIds = modelIdAndNodeIds.get(e).toArray(new String[] {});
+                    log.error("fail to deploy model {} under the node ids: {} \nthe reason is: {}", e, nodeIds, exception);
+
+                    // store the latest value of the retryTimes and node id under the
+                    // index ".plugins-ml-model-redeploy"
+                    saveLatestDeployRetryTimes(
+                        localNodeId,
+                        finalRetryTimes,
+                        ActionListener
+                            .wrap(
+                                saveLatestDeployRetryTimesStepWhenThrowException::onResponse,
+                                saveLatestDeployRetryTimesStepWhenThrowException::onFailure
+                            )
+                    );
+
+                    saveLatestDeployRetryTimesStepWhenThrowException
+                        .whenComplete(
+                            response -> log
+                                .info("insert retry times successfully before throwing exception, retryTimes is {}", finalRetryTimes),
+                            saveLatestDeployRetryTimesStep::onFailure
+                        );
+
+                    throw new MLException(exception);
+                }))
+                .collect(Collectors.toList());
+
+            for (int i = 0; i < modelIds.length; i++) {
+                String[] nodeIds = new ArrayList<>(modelIdAndNodeIds.get(modelIds[i])).toArray(new String[] {});
+
+                ActionListener<MLDeployModelResponse> currentListener = listeners.get(i);
+                autoDeployModelByNodeAndModelId(nodeIds, modelIds[i], currentListener);
```

Comment:
This deploy method is async, it will return task id directly, so that means we will deploy multiple models in parallel?

Reply:
The previous code looked like this
 ```
for (int i = 0; i < modelIds.length; i++) {
                String currentModelId = modelIds[i].
                ActionListener<MLDeployModelResponse> listener = listeners.get(i).
                nodeIds = new ArrayList<>(modelIdAndNodeIds.get(currentModelId)).toArray(new String[] {}).

                if (i == 0) {
                    autoDeployModelByNodeAndModelId(nodeIds, currentModelId, listener).
                }

                if (i + 1 < modelIds.length) {
                    String nextModelId = modelIds[i + 1].
                    nodeIds = new ArrayList<>(modelIdAndNodeIds.get(nextModelId)).toArray(new String[] {}).
                    ActionListener<MLDeployModelResponse> listenerNext = listeners.get(i + 1).
                    autoDeployModelByNodeAndModelId(nodeIds, nextModelId, listenerNext).
                }
            }

```
I think it can be optimized to the current code, so it's now like this.

Reply:
@ylwu-amzn  I used `StepListener` class to let model deployed one by one, u can see the latest code